h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -329,9 +345,13 @@ private void gridSearch(Grid<MP> grid) {
     try {
       // Get iterator to traverse hyper space
       HyperSpaceWalker.HyperSpaceIterator<MP> it = _hyperSpaceWalker.iterator();
-      // Number of traversed model parameters
-      int counter = grid.getModelCount();
-      while (it.hasNext(model)) {
+
+      // This counter will be used as index to generate keys for corresponding `successful` models
+      int numberOfBuiltModels = grid.getModelCount();
+
+      int maxNumberOfModelsToBuild = numberOfBuiltModels + it.max_models();
```

Comment:
I do not understand this line. For example for the Cartesian walker, it.max_models is the size of the hyperspace. Adding any positive number to it results in a larger size than the size of the hyperspace. I am not sure what the `maxNumberOfModelsToBuild` really represents then.

Reply:
@michalkurka please see my explanation here https://github.com/h2oai/h2o-3/pull/4191#discussion_r370112796